File: apps/java/design-pattern/src/test/java/pattern/templatemethod/TemplateMethodTest.java
package pattern.templatemethod;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class TemplateMethodTest {

    @Test
    void htmlReportOutputsValidHtml() {
        Report report = new HtmlReport();
        String output = report.outputReport();

        assertTrue(output.startsWith("<html>"));
        assertTrue(output.endsWith("</html>\n"));
        assertTrue(output.contains("<title>月次報告</title>"));
        assertTrue(output.contains("<p>順調</p>"));
        assertTrue(output.contains("<p>最高の調子</p>"));
    }

    @Test
    void plainTextReportOutputsFormattedText() {
        Report report = new PlainTextReport();
        String output = report.outputReport();

        assertTrue(output.startsWith("**** 月次報告 ****"));
        assertTrue(output.contains("順調"));
        assertTrue(output.contains("最高の調子"));
        assertFalse(output.contains("<html>"));
    }

    @Test
    void htmlReportContainsBodyTags() {
        Report report = new HtmlReport();
        String output = report.outputReport();

        assertTrue(output.contains("<body>"));
        assertTrue(output.contains("</body>"));
    }

    @Test
    void plainTextReportDoesNotContainHtmlTags() {
        Report report = new PlainTextReport();
        String output = report.outputReport();

        assertFalse(output.contains("<html>"));
        assertFalse(output.contains("<body>"));
        assertFalse(output.contains("<p>"));
    }
}

File: apps/ruby/design-pattern/04-template-method/test/template_method_test.rb
# frozen_string_literal: true

require_relative '../../test/test_helper'
require_relative '../lib/html_report'
require_relative '../lib/plain_text_report'

class TemplateMethodTest < Minitest::Test
  def test_html_report_output
    report = TemplateMethod::HtmlReport.new

    expected = <<~HTML
      <html>
       <head>
       <title>月次報告</title>
       </head>
      <body>
       <p>順調</p>
       <p>最高の調子</p>
      </body>
      </html>
    HTML

    assert_output(expected) { report.output_report }
  end

  def test_plain_text_report_output
    report = TemplateMethod::PlainTextReport.new

    expected = <<~TEXT
      **** 月次報告 ****

      順調
      最高の調子
    TEXT

    assert_output(expected) { report.output_report }
  end

  def test_base_report_raises_on_output_line
    report = TemplateMethod::Report.new

    assert_raises(NotImplementedError) { report.output_report }
  end
end
